File: problem/Longest_Common_Prefix/longestCommonPrefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        Vertical scanning approach - O(S) time, O(1) space
        where S is the sum of all characters in all strings
        """
        if not strs:
            return ""
        
        min_len = min(len(s) for s in strs)
        
        for i in range(min_len):
            char = strs[0][i]

            for j in range(1, len(strs)):
                if strs[j][i] != char:
                    return strs[0][:i]
        
        return strs[0][:min_len]
    
    def longestCommonPrefix_horizontal(self, strs):
        """
        Horizontal scanning approach
        """
        if not strs:
            return ""
        
        prefix = strs[0]
        
        for i in range(1, len(strs)):
            j = 0
            while j < len(prefix) and j < len(strs[i]) and prefix[j] == strs[i][j]:
                j += 1
            prefix = prefix[:j]

            if not prefix:
                return ""
        
        return prefix
    
    def longestCommonPrefix_divide_conquer(self, strs):
        """
        Divide and conquer approach
        """
        def common_prefix(left, right):
            min_len = min(len(left), len(right))
            for i in range(min_len):
                if left[i] != right[i]:
                    return left[:i]
            return left[:min_len]
        
        def divide_conquer(strs, l, r):
            if l == r:
                return strs[l]
            
            mid = (l + r) // 2
            left_prefix = divide_conquer(strs, l, mid)
            right_prefix = divide_conquer(strs, mid + 1, r)
            
            return common_prefix(left_prefix, right_prefix)
        
        if not strs:
            return ""
        
        return divide_conquer(strs, 0, len(strs) - 1)

File: problem/Longest_Common_Prefix/longestCommonPrefix.c
char *longestCommonPrefix(char **strs, int strsSize)
{
    if (strsSize == 0)
    {
        char *result = (char *)malloc(1);
        result[0] = '\0';
        return result;
    }

    int minLen = strlen(strs[0]);
    for (int i = 1; i < strsSize; i++)
    {
        int len = strlen(strs[i]);
        if (len < minLen)
        {
            minLen = len;
        }
    }

    char *result = (char *)malloc((minLen + 1) * sizeof(char));

    for (int i = 0; i < minLen; i++)
    {
        char c = strs[0][i];

        for (int j = 1; j < strsSize; j++)
        {
            if (strs[j][i] != c)
            {
                result[i] = '\0';
                return result;
            }
        }

        result[i] = c;
    }

    result[minLen] = '\0';
    return result;
}

char *longestCommonPrefix_horizontal(char **strs, int strsSize)
{
    if (strsSize == 0)
    {
        char *result = (char *)malloc(1);
        result[0] = '\0';
        return result;
    }

    int prefixLen = strlen(strs[0]);
    char *result = (char *)malloc((prefixLen + 1) * sizeof(char));
    strcpy(result, strs[0]);

    for (int i = 1; i < strsSize; i++)
    {
        int j = 0;
        while (j < prefixLen && j < strlen(strs[i]) &&
               result[j] == strs[i][j])
        {
            j++;
        }
        result[j] = '\0';
        prefixLen = j;

        if (prefixLen == 0)
            break;
    }

    return result;
}
